Framework/DataObjects/test/TimeSplitterTest.h

```diff
@@ -180,4 +183,67 @@ class TimeSplitterTest : public CxxTest::TestSuite {
     TS_ASSERT(!roi.empty());
     TS_ASSERT_EQUALS(roi.numBoundaries(), 4);
   }
+
+  void test_timeSplitterFromMatrixWorkspace() {
+    TimeSplitter splitter;
+    splitter.addROI(DateAndTime(0, 0), DateAndTime(10, 0), 1);
+    splitter.addROI(DateAndTime(10, 0), DateAndTime(15, 0), 3);
+    splitter.addROI(DateAndTime(15, 0), DateAndTime(20, 0), 2);
+
+    Mantid::API::MatrixWorkspace_sptr ws = WorkspaceCreationHelper::create2DWorkspaceBinned(1, 3);
+
+    auto &X = ws->dataX(0);
+    auto &Y = ws->dataY(0);
+    // X[0] is 0 by default, unit is seconds.
```

Comment:
@gecage952 Did you mean `unit is nanoseconds`? In the `TimeSplitter` constructor you cast to `int64_t`, so I think it is nanoseconds.

Reply:
Actually we need to make sure what the time units of the input MatrixWorkspace are.

Reply:
[Here it is in the code](https://github.com/mantidproject/mantid/blob/main/Framework/Algorithms/src/FilterEvents.cpp#L892):
```cpp
  // Transform vector
  for (size_t i = 0; i < sizex; ++i) {
    m_vecSplitterTime[i] = static_cast<int64_t>(X[i] * 1.E9);
  }
```
So the units of `X` are _seconds_.
@gecage952 can you amend the docs to specify that units are seconds in case of `MatrixWorkspace`?

Reply:
Yeah I can do that. should I do so in this pr or a new one?

Reply:
@jmborr, My bad, I thought if it's `int64_t`, it must be nanoseconds.

@gecage952, Well, if it is seconds, then I am concerned about this code:
```
auto timeStart = Types::Core::DateAndTime(static_cast<int64_t>((X[i - 1])), offset.totalNanoseconds()
```
Can you test it with a huge offset like today's date - see ONE,TWO,THREE... in the test file.  As I mentioned in another comment, I had to create a `DateAndTime` object without the offset first, and then add the offset separately.  Otherwise it just didn't work with "huge" offsets.

Reply:
you just need to multiply `X[i - 1]` by `secTonanosec`.
same for `X[i]`

Reply:
Of course, in my case the constructor was `DateAndTime(const double seconds, const double nanoseconds)` while in your case it is `DateAndTime(const int64_t seconds, const int64_t nanoseconds)`, so it might work - I am not sure.  But, in any event, I think using ONE,TWO,or THREE as an offset is a more realistic test, because for a user the offset most likely will be the start of an experiment run.

Reply:
```c++
/** Construct a time from the number of seconds and nanoseconds since Jan 1,
 * 1990.
 * @param seconds :: seconds since Jan 1, 1990.
 * @param nanoseconds :: nanoseconds to add to the number of seconds
 */
DateAndTime::DateAndTime(const int64_t seconds, const int64_t nanoseconds) {
  if (seconds >= MAX_SECONDS)
    _nanoseconds = MAX_NANOSECONDS;
  else if (seconds <= MIN_SECONDS)
    _nanoseconds = MIN_NANOSECONDS;
  else
    _nanoseconds = seconds * NANO_PER_SEC + nanoseconds;
}
```

This is the constructor in DateAndTime that I'm using. I thought it kinda did that already

Reply:
Holy crap, I haven't realized the constructor for `DateAndTime` mixed seconds and nanoseconds. What a stupid thing to do :angry:

Reply:
@jmborr I think they tried to make user's (not programmer's) life easier 😃

Reply:
not only stupid but dangerous!! For instance, if I have:
```
double X[i] = 42.5;
// invoke constructor DateAndTime::DateAndTime(const int64_t seconds, const int64_t nanoseconds)
DateAndTime(static_cast<int64_t>(42.5), 12345567);
```
Statement `static_cast<int64_t>(42.5)` will cutoff `0.5` seconds, or `500000000` nanoseconds!!!

Reply:
we should be using `DateAndTime(const double seconds, const double nanoseconds)` here too

Reply:
Well, what worked for me was: `DateAndTime(myseconds, 0.0)` and then add the offset separately.  Otherwise, when I tried passing both seconds and nanoseconds to the constructor, it didn't work if nanoseconds were like ONE,TWO,THREE.

Reply:
Ok, I'll do the add then.